added nix development environment, Cargo.lock, edition update, dependency update

Guelakais opened the pull request:
the added `flake.nix` is there to be able to build this package in nix environments.
A `Cargo.lock` file is urgently required for certain packet managers. Therefore, the corresponding entry in the `.gitignore` has been removed to enable its generation.

Guelakais committed:
added nix development environment

added rust-toolchain.toml for the nix flake

The nix flake needs a rust-toolchain.toml to rely on

Guelakais committed:
added cargo.lock

esteve commented:
@Guelakais please avoid adding any changes that are unrelated to the goal of this PR. You mentioned that you're reenabling `Cargo.lock`, but at the same time you've added changes to the source code and to the toolchain, please revert those.

I've also seen the related link to the nixpkgs project, my suggestion in general is that you should listen more to the maintainers of a project and not take feedback personally, not everyone has as much patience as we have demonstrated multiple times.

Guelakais commented:
I have noticed that the builds always fail because your toolchain for maturin is outdated. My solution to this is to switch directly to the `nightly` toolchain, as this error should then no longer occur.

Guelakais committed:
applying rust warnings in `main.rs`

```rust
warning: the borrowed expression implements the required traits
   --> src/lib.rs:149:23
    |
149 |         std::fs::copy(&src, &dest).with_context(|| {
    |                       ^^^^ help: change this to: `src`
    |
    = help: for further information visit
https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrows_for_generic_args
    = note: `#[warn(clippy::needless_borrows_for_generic_args)]` on by
default

warning: `cargo-ament-build` (lib) generated 1 warning (run `cargo
clippy --fix --lib -p cargo-ament-build` to apply 1 suggestion)
```

Guelakais committed:
edition update

Guelakais committed:
dependency update

esteve commented:
Closing this PR as the feedback hasn't been addressed and it hasn't been updated in a long time. @Guelakais please reopen or submit a new PR when you have addressed the feedback I gave you, thanks.
